### PR TITLE
[Frozen Period] Special Release, March 2022

### PR DESCRIPTION
**Pull request by nariman87**
## Context for changes

- **Upload workflows have been tested, updated, and executed**: 
    `__version__` was updated to reflect recent changes, upload workflows have been tested, and a release tag will be issued after this PR. Note, only OWNERS can manually trigger the upload workflows.
- **The codefactor quality score has been brought back to `B-`**: 
    Temporarily relaxed the required standards for codefactor quality score to a `B-` (see https://github.com/XanaduAI/ft-stack/blob/main/.github/pull_request_template.md) or better. We still have a large number of migrated codes that need quality improvements in near future PRs. Note `B-` still has a green-highlighted badge. 

## Example usage and tests

- [x] Not Applicable

## Performance results justifying changes

- [x] Not Applicable

## Workflow actions and tests

- [x] Not Applicable

## Expected benefits and drawbacks

**Expected benefits:**
- Preparing the repo and PyPI packages for the new tagged release. 

## Related Github issues

- [x] Not Applicable

## Checklist and integration statements

- [x] My Python and C++ codes follow the coding and commenting styles of this project as indicated by existing files. Specifically, the changes conform to given `black`, `docformatter` and `pylint` configurations. 
- [x] I have performed a self-review of these changes. I have checked my code and corrected misspellings to the best of my capacity. I have checked the [CodeFactor score](https://www.codefactor.io/repository/github/xanaduai/flamingpy/branches) in the active branch and ensured it is `B` or higher. I also confirm that I have already merged other branches into this branch as required.
- [x] I have added context for corresponding changes in documentation and README.md as needed.
- [x] I have added new workflow CI tests for corresponding changes and these pass locally for me.
- [x] I have updated `_version.py` based on [semantic versioning](https://semver.org/). I recognize that the developers may create a Special Release including my changes.

### CODE REVIEW

File: .github/workflows/format.yaml [hunk 1]
```diff
@@ -1,13 +1,15 @@
 name: Formatting Checks
 on:
-  pull_request:
+  push:
     branches:
       - main
       - develop
-  push:
+      - sims2cloud
```
Comment (ilan-tz) [inline .github/workflows/format.yaml:7]: Do we anticipate having a public sims2cloud branch?

File: .github/workflows/format.yaml [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
+  pull_request:
     branches:
       - main
-      - develop    
+      - develop
+      - sims2cloud
```
Comment (ilan-tz) [inline .github/workflows/format.yaml:12]: Same comment as above.

File: .github/workflows/linting.yaml [hunk 3]
```diff
@@ -1,9 +1,15 @@
 name: Linting Checks
 on:
+  push:
+    branches:
+      - main
+      - develop
+      - sims2cloud
```
Comment (ilan-tz) [inline .github/workflows/linting.yaml:7]: Same comment as above.

File: .github/workflows/linting.yaml [hunk 4]
```diff
@@ -1,9 +1,15 @@
...
   pull_request:
-    branches:    
+    branches:
       - main
       - develop
+      - sims2cloud
```
Comment (ilan-tz) [inline .github/workflows/linting.yaml:12]: Same comment as above.

File: doc/custom_directives.py [hunk 5]
```diff
@@ -87,6 +86,7 @@ class CustomGalleryItemDirective(Directive):
     add_index = False
 
     def run(self):
+        """An auxiliary function to initiate creating the thumbnail"""
```
Comment (ilan-tz) [inline doc/custom_directives.py:89]:
```suggestion
        """An auxiliary function to initiate creating the thumbnail."""
```

File: doc/custom_directives.py [hunk 6]
```diff
@@ -155,6 +155,7 @@ class DetailsDirective(Directive):
     add_index = False
 
     def run(self):
+        """An auxiliary function to initiate creating the section"""
```
Comment (ilan-tz) [inline doc/custom_directives.py:158]:
```suggestion
        """An auxiliary function to initiate creating the section."""
```

File: flamingpy/benchmarks/simulations.py [hunk 7]
```diff
@@ -54,7 +54,7 @@ def ec_monte_carlo(code, trials, delta, p_swap, passive_objects=None, backend="c
     """
     cv_noise = {"noise": "grn", "delta": delta, "sampling_order": "initial"}
     if passive_objects is not None:
-        RHG_macro, RHG_reduced, CVRHG_reduced, bs_network = passive_objects
+        # RHG_macro, RHG_reduced, CVRHG_reduced, bs_network = passive_objects
```
Comment (ilan-tz) [inline flamingpy/benchmarks/simulations.py:57]:
```suggestion
```

File: flamingpy/simulations.py [hunk 8]
```diff
@@ -44,7 +44,7 @@ def ec_monte_carlo(code, trials, delta, p_swap, passive_objects=None):
         errors (integer): the number of errors.
     """
     if passive_objects is not None:
-        RHG_macro, RHG_reduced, CVRHG_reduced, bs_network = passive_objects
+        # RHG_macro, RHG_reduced, CVRHG_reduced, bs_network = passive_objects
```
Comment (ilan-tz) [inline flamingpy/simulations.py:47]:
```suggestion
```

File: setup.py [hunk 9]
```diff
@@ -27,26 +28,34 @@
...
 # Reading the package version number
-with open("flamingpy/_version.py") as f:
+with open("flamingpy/_version.py", encoding="utf8") as f:
     version = f.readlines()[-1].split()[-1].strip("\"'")
 
 
 class BinaryDistribution(dist.Distribution):
+    """A class to define Binary Distribution objects"""
     def has_ext_modules(foo):
+        """Checking for external modules"""
```
Comment (ilan-tz) [inline setup.py:38]:
```suggestion
        """Check for external modules."""
```

File: setup.py [hunk 10]
```diff
@@ -27,26 +28,34 @@
...
         return True
 
 
-# The following class is an adaptation of Python examples for pybind11:
-# https://github.com/pybind/python_example/blob/master/setup.py
 class CMakeExtension(Extension):
+    """A class to define CMake Extensions.
+    
+    The following class is an adaptation of Python examples for pybind11:
+    https://github.com/pybind/python_example/blob/master/setup.py
```
Comment (ilan-tz) [inline setup.py:46]:
```suggestion
    Adapted from Python examples for pybind11:
    https://github.com/pybind/python_example/blob/master/setup.py
```

File: setup.py [hunk 11]
```diff
@@ -27,26 +28,34 @@
...
         self.sourcedir = os.path.abspath(sourcedir)
 
 
-# The following class was adapted from pymatching package:
-# https://github.com/oscarhiggott/PyMatching/blob/master/setup.py
 class CMakeBuild(build_ext):
+    """A class to define, configure, and test build extensions.
+    
+    The following class was adapted from pymatching package:
+    https://github.com/oscarhiggott/PyMatching/blob/master/setup.py
```
Comment (ilan-tz) [inline setup.py:57]:
```suggestion
    Adapted from the pymatching package:
    https://github.com/oscarhiggott/PyMatching/blob/master/setup.py
```

File: tests/test_benchmarks_examples.py [hunk 12]
```diff
@@ -16,40 +16,50 @@
 
 
 def test_decoder_example():
-    from flamingpy.examples import decoding
+    """Simple tests for decoding module in flamingpy.examples"""
```
Comment (ilan-tz) [inline tests/test_benchmarks_examples.py:19]:
```suggestion
    """Simple test for the decoding module in flamingpy.examples."""
```

File: tests/test_benchmarks_examples.py [hunk 13]
```diff
@@ -16,40 +16,50 @@
...
+    from flamingpy.examples import decoding as dc_examples
+
+
+def test_decoding_benchmark():
+    """Simple tests for decoding module in flamingpy.benchmarks"""
```
Comment (ilan-tz) [inline tests/test_benchmarks_examples.py:24]:
```suggestion
    """Simple test for the decoding module in flamingpy.benchmarks."""
```

File: tests/test_benchmarks_examples.py [hunk 14]
```diff
@@ -16,40 +16,50 @@
...
+    from flamingpy.benchmarks import decoding as dc_benchmarks
 
 
 def test_gkp_example():
+    """Simple tests for gkp module in flamingpy.examples"""
```
Comment (ilan-tz) [inline tests/test_benchmarks_examples.py:29]:
```suggestion
    """Simple test for the gkp module in flamingpy.examples."""
```

File: tests/test_benchmarks_examples.py [hunk 15]
```diff
@@ -16,40 +16,50 @@
...
     from flamingpy.examples import gkp
 
 
 def test_graphstates_example():
+    """Simple tests for graphstates module in flamingpy.examples"""
```
Comment (ilan-tz) [inline tests/test_benchmarks_examples.py:34]:
```suggestion
    """Simple test for the graphstates module in flamingpy.examples."""
```

File: tests/test_benchmarks_examples.py [hunk 16]
```diff
@@ -16,40 +16,50 @@
...
     from flamingpy.examples import graphstates
 
 
 def test_macro_reduce_example():
+    """Simple tests for macro_reduce module in flamingpy.examples"""
```
Comment (ilan-tz) [inline tests/test_benchmarks_examples.py:39]:
```suggestion
    """Simple test for the macro_reduce module in flamingpy.examples."""
```

File: tests/test_benchmarks_examples.py [hunk 17]
```diff
@@ -16,40 +16,50 @@
...
     from flamingpy.examples import macro_reduce
 
 
 def test_surface_code_example():
+    """Simple tests for surface_code module in flamingpy.examples"""
```
Comment (ilan-tz) [inline tests/test_benchmarks_examples.py:44]:
```suggestion
    """Simple test for the surface_code module in flamingpy.examples."""
```

File: tests/test_benchmarks_examples.py [hunk 18]
```diff
@@ -16,40 +16,50 @@
...
     from flamingpy.examples import surface_code
 
 
-def test_decoding_benchmark():
-    from flamingpy.benchmarks import decoding
-
-
 def test_lemon_benchmark():
+    """Simple tests for lemon module in flamingpy.benchmarks"""
```
Comment (ilan-tz) [inline tests/test_benchmarks_examples.py:49]:
```suggestion
    """Simple test for the lemon module in flamingpy.benchmarks."""
```

File: tests/test_benchmarks_examples.py [hunk 19]
```diff
@@ -16,40 +16,50 @@
...
     from flamingpy.benchmarks import lemon
 
 
 def test_matching_benchmark():
+    """Simple tests for matching module in flamingpy.benchmarks"""
```
Comment (ilan-tz) [inline tests/test_benchmarks_examples.py:54]:
```suggestion
    """Simple test for the matching module in flamingpy.benchmarks."""
```

File: tests/test_benchmarks_examples.py [hunk 20]
```diff
@@ -16,40 +16,50 @@
...
     from flamingpy.benchmarks import matching
 
 
 def test_shortest_path_benchmark():
+    """Simple tests for shortest_path module in flamingpy.benchmarks"""
```
Comment (ilan-tz) [inline tests/test_benchmarks_examples.py:59]:
```suggestion
    """Simple test for the shortest_path module in flamingpy.benchmarks."""
```

File: tests/test_benchmarks_examples.py [hunk 21]
```diff
@@ -16,40 +16,50 @@
...
     from flamingpy.benchmarks import shortest_path
 
 
 def test_simulations_benchmark():
+    """Simple tests for simulations module in flamingpy.benchmarks"""
```
Comment (ilan-tz) [inline tests/test_benchmarks_examples.py:64]:
```suggestion
    """Simple test for the simulations module in flamingpy.benchmarks."""
```

File: tests/unit/test_decoder.py [hunk 22]
```diff
@@ -136,7 +136,7 @@ def test_stab_graph(self, enc_state):
 
     def test_matching_graph(self, match_data):
         """Test the structure of the matching graph."""
-        for i, graph in enumerate(match_data[0]):
+        for _, graph in enumerate(match_data[0]):
```
Comment (ilan-tz) [inline tests/unit/test_decoder.py:139]:
```suggestion
        for graph in match_data[0]:
```

**Review by ilan-tz — CHANGES_REQUESTED** [review]
Great work! Just a few changes and questions.

File: .github/workflows/format.yaml (hunk 1 above)
Comment (nariman87) [reply .github/workflows/format.yaml:7]: Not really, this is meant for the private project! One should keep in mind it is harmless to add private branches here. But when this is synced with `main` of ft-stack, and then people sync `sims2cloud` with that main, it means tests will be properly executed on the private branch.

File: .github/workflows/format.yaml (hunk 2 above)
Comment (nariman87) [reply .github/workflows/format.yaml:12]: See if my comment above makes sense.

File: .github/workflows/linting.yaml (hunk 3 above)
Comment (nariman87) [reply .github/workflows/linting.yaml:7]: See if my comment above makes sense.

File: .github/workflows/linting.yaml (hunk 4 above)
Comment (nariman87) [reply .github/workflows/linting.yaml:12]: See if my comment above makes sense.

**Comment by ilan-tz** [comment]
Let us please also add cpp_mc_loop.cpp to .gitignore

**Comment by nariman87** [comment]
> Let us please also add cpp_mc_loop.cpp to .gitignore

Done.

**Review by ilan-tz — APPROVED** [review]

**Pull request merged**